### PR TITLE
Fix issue with Win32 resource placement

### PR DESCRIPTION
**Pull request by davidwrighton**
Windows has a bug in its resource processing logic that occurs when
1. A PE file is loaded as a data file
2. The resource data found in the resources has an RVA which has a magnitude greater than the size of the section which holds the resources
3. The offset of the start of the resource data from the start of the section is not zero.

As it is impossible to effect condition 1 in the compiler, and changing condition 2 would require bloating the virtual size of the sections, instead require that the resource data is located at offset 0 within its section, by sorting the Win32ResourceNode to be the first node. (This issue caused issues while developing dotnet/aspnetcore#31778.)

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib

**Review by mangod9 — APPROVED** [review]

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/SectionBuilder.cs [hunk 1]
```diff
@@ -834,6 +834,15 @@ public void UpdateDirectories(PEDirectoriesBuilder directoriesBuilder)
...
                 Section section = _sections[symbolTarget.SectionIndex];
                 Debug.Assert(section.RVAWhenPlaced != 0);
+
+                // Windows has a bug in its resource processing logic that occurs when 
+                // 1. A PE file is loaded as a data file
+                // 2. The resource data found in the resources has an RVA which has a magnitude greater than the size of the section which holds the resources
+                // 3. The offset of the start of the resource data from the start of the section is not zero.
+                //
+                // As it is impossible to effect condition 1 in the compiler, and changing condition 2 would require bloating the virtual size of the sections,
+                // instead require that the resource data is located at offset 0 within the section.
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/SectionBuilder.cs:844]: Might save someone's time when this assert fires:

```suggestion
                // instead require that the resource data is located at offset 0 within the section.
                // We achieve that by sorting the Win32ResourcesNode as the first node.
```

**Review by MichalStrehovsky — APPROVED** [review]
I guess this is why most compilers place resources in a dedicated `.rsrc` section instead of `.text`? We could probably also do that so that we don't clutter code pages with rarely used data.

**Review by trylek — APPROVED** [review]
Looks great, thanks for fixing this!

**Comment by davidwrighton** [comment]
@MichalStrehovsky Putting stuff in the rsrc section makes sense, but it would cause a file size increase for very small binaries such as the problematic one, and we have quite a few small little binaries.

**Pull request merged**